Automattic/woocommerce-payments
# Fix Recurring Product Discount coupon double conversion

Fixes #2354

#### Changes proposed in this Pull Request

* Create new `should_convert_coupon_amount` that will check to see if the coupon amount should be converted.
* This new method checks to see if there is a renewal in the cart, and then checks for certain actions/calls in the backtrace in order to move forward. From there, it checks to see if the coupon is a recurring type coupon and will determine if it should return `false`.
* New method added to `FrontendPrices->get_coupon_amount()` to actually check to see if conversion is needed.
* Tests added/updated as needed.

#### Testing instructions

1. Make sure you have _Early Renewals_ enabled under WooCommerce > Settings > Subscriptions.
1. Create a _Recurring Product Discount_ coupon that is less than the amount of the product you're going to use for testing.
1. Purchase the subscription product with the coupon.
2. Go to My Account > Subscriptions, and then that subscription.
3. Click _Renew now_, and you will see the coupon amount has not been converted again.

Secondary test:

1. Navigate to WooCommerce > Subscriptions in the admin.
2. Select the subscription from above.
3. Under _Subscriptions actions_ in the top, right corner, select _Create pending renewal order_ and then click the little arrow or _Update_.
4. A new order will be created, and a link to it will be in the _Subscription notes_, click the link to go to that order.
5. There will be a _Customer payment page_ link above the _Status_ select box, click that.
6. Verify that the coupon amount has not been converted a second time. 

-------------------
